src/nni_evaluation_engine.hpp

```diff
@@ -0,0 +1,184 @@
+  const GraftDAG *graft_dag_ = nullptr;
+  // Scored NNIs.
+  NNIDoubleMap scored_nnis_;
+};
+
+class NNIEvaluationEngineViaGP : public NNIEvaluationEngine {
+ public:
+  NNIEvaluationEngineViaGP(NNIEngine &nni_engine, GPEngine &gp_engine);
+
+  // Initialize Evaluation Engine.
+  virtual void Init();
+  // Prepare Engine for NNI Engine loop.
+  virtual void Prep();
+  // Resize Engine for modified DAG.
+  virtual void GrowEngineForDAG(std::optional<Reindexer> node_reindexer,
+                                std::optional<Reindexer> edge_reindexer);
+  // Fetches Pre-NNI data to prep Post-NNI for likelihood computation. Method stores
+  // intermediate values in the GPEngine temp space (expects GPEngine has already been
+  // resized).
+  virtual void GrowEngineForAdjacentNNIs(const NNISet &adjacent_nnis,
+                                         const bool via_reference = true,
+                                         const bool use_unique_temps = true);
+  // Compute scores for all NNIs adjacent to current DAG.
+  virtual void ScoreAdjacentNNIs(const NNISet &adjacent_nnis);
+
+  // ** Helpers
+
+  // Build GPOperation vector for computing NNI Likelihood. For NNIs that are not in
+  // the DAG, must create KeyIndexMap through
+  // PassGPEngineDataFromPreNNIToPostNNIViaReference.
+  GPOperationVector BuildGPOperationsForNNILikelihood(
+      const NNIOperation &nni, const KeyIndexMap &nni_key_idx) const;
+  // Build GPOperation vector for computing NNI Likelihood for all adjacent NNIs.
+  GPOperationVector BuildGPOperationsForAdjacentNNILikelihoods(
+      const NNISet &adjacent_nnis, const bool via_reference);
+  // Grow engine to handle computing NNI Likelihoods for all adjacent NNIs.
+  // Option to grow engine for computing likelihoods via reference or via copy. If
+  // computing via reference, option whether to use unique temporaries (for testing and
+  // computing in parallel).
+  void GrowGPEngineForAdjacentNNILikelihoods(const NNISet &adjacent_nnis,
+                                             const bool via_reference = true,
+                                             const bool use_unique_temps = true);
+  // Fetches Data from Pre-NNI for Post-NNI. Can be used for initial values when moving
+  // accepted NNIs from Graft to Host DAG.
+  KeyIndexMapPair PassGPEngineDataFromPreNNIToPostNNIViaCopy(
+      const NNIOperation &pre_nni, const NNIOperation &post_nni);
+  // Fetches Pre-NNI data to prep Post-NNI for likelihood computation. Method stores
+  // intermediate values in the GPEngine temp space (expects GPEngine has already been
+  // resized).
+  KeyIndexMapPair PassGPEngineDataFromPreNNIToPostNNIViaReference(
+      const NNIOperation &pre_nni, const NNIOperation &post_nni, const size_t nni_count,
+      const bool use_unique_temps);
+
+  // ** Access
+
+  GPEngine &GetGPEngine() { return *gp_engine_; }
+
+ protected:
+  GPEngine *gp_engine_ = nullptr;
+};
+
+class NNIEvaluationEngineViaTP : public NNIEvaluationEngine {
```

Comment:
I'm confused about the role of this versus the TPEngine. Would it be helpful to have an overview somewhere? There's a lot here!

Reply:
I tried to add more description to the headers of the TPEngine and TPEvalEngine.  Basically the main thing that TPEngine does outside of the TPEvalEngine is to maintain the choice map, which contains the topologies for the top trees indexed by its focal edge.

Reply:
The rest is mostly passthrough calls to the TPEvalEngine.